src/services/bunker.py

```diff
@@ -1,15 +1,344 @@
+
+NORMALIZED_CL_PER_EPOCH = 64
+MISTAKE_RANGE = 0.1
+NEAREST_EPOCH_DISTANCE = 4
+FAR_EPOCH_DISTANCE = 25
+
+logger = logging.getLogger(__name__)
+
 
 class BunkerService:
-    def __init__(self, w3: Web3):
+
+    def __init__(self, w3: Web3, frame_config: FrameConfig, chain_config: ChainConfig):
         self.w3 = w3
+        self.cm = ConsensusModule(self.w3)
+        self.f_conf = frame_config
+        self.c_conf = chain_config
+
+    def is_bunker_mode(self, cdtp: CommonDataToProcess) -> bool:
+        logger.info({"msg": "Checking bunker mode"})
+        frame_cl_rebase = self._get_cl_rebase_for_frame(cdtp)
+        if frame_cl_rebase < 0:
+            logger.info({"msg": "Bunker ON. CL rebase is negative"})
+            return True
+        if self._is_high_midterm_slashing_penalty(cdtp, frame_cl_rebase):
+            logger.info({"msg": "Bunker ON. High midterm slashing penalty"})
+            return True
+        if self._is_abnormal_cl_rebase(cdtp, frame_cl_rebase):
+            logger.info({"msg": "Bunker ON. Abnormal CL rebase"})
+            return True
+
+    def _get_cl_rebase_for_frame(self, cdtp: CommonDataToProcess) -> Gwei:
+        logger.info({"msg": "Getting CL rebase for frame"})
+        ref_lido_balance, ref_withdrawal_vault_balance = self._calculate_lido_balance_with_vault(
+            cdtp.ref_lido_validators, cdtp.ref_blockstamp
+        )
+        # Make a static call of 'handleOracleReport' without EL rewards
+        # to simulate report and get total pool ether after that
+        args = {
+            "_reportTimestamp": cdtp.ref_timestamp,
+            "_timeElapsed": cdtp.seconds_elapsed_since_last_report,
+            "_clValidators": len(cdtp.ref_lido_validators),
+            "_clBalance": self.w3.to_wei(ref_lido_balance, 'gwei'),
+            "_withdrawalVaultBalance": ref_withdrawal_vault_balance,
+            "_elRewardsVaultBalance": 0,
+            "_lastFinalizableRequestId": 0,
+            "_simulatedShareRate": 0,
+        }
+        before_report_total_pooled_ether = self.w3.lido_contracts.lido.functions.totalSupply().call(
+            block_identifier=cdtp.ref_blockstamp.block_hash
+        )
+        after_report_total_pooled_ether, _ = self.w3.lido_contracts.lido.functions.handleOracleReport(**args).call(
+            block_identifier=cdtp.ref_blockstamp.block_hash
+        )
+        logger.info({
+            "msg": "Simulate 'handleOracleReport' contract function call",
+            "call_args": args,
+            "before_call": before_report_total_pooled_ether,
+            "after_call": after_report_total_pooled_ether,
+        })
+
+        frame_cl_rebase = self.w3.from_wei(after_report_total_pooled_ether - before_report_total_pooled_ether, 'gwei')
+        logger.info({"msg": f"Simulated CL rebase for frame: {frame_cl_rebase} Gwei"})
+
+        return frame_cl_rebase
+
+    def _is_high_midterm_slashing_penalty(self, cdtp: CommonDataToProcess, cl_rebase: Gwei) -> bool:
```

Comment:
Lets split this function. It should increase readability and decrease tests difficalty

Reply:
Yes. The major part should be static. will be done